lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp

```diff
@@ -866,14 +869,23 @@ static void fuseZeroWithGemmOrBrgemm(RewriterBase &rewriter,
   if (it == itEnd)
     return;
   while (++it != itEnd) {
-    if (mlir::isPure(&*it))
+    // Skip pure operations and operations that do not
+    // touch `dest`.
```

Comment:
Not sure if that's safe in case of aliasing.

Reply:
if we have a subview on `dest` we fails.

Reply:
Do you have in mind a specific pattern?

Reply:
> if we have a subview on dest we fails.

Why would we, it is pure afaik and touches dest.

For example:
```
func.func @zero_test(%arg0: memref<1x32x32xf32>, %arg1: memref<1x32x32xf32>, %arg2: memref<32x32xf32>) {
  %c1_i64 = arith.constant 1 : i64
  %cst = arith.constant 0.000000e+00 : f32
  %alloc = memref.alloc() {alignment = 64 : i64} : memref<32x32xf32>
  %0 = xsmm.unary.dispatch zero [32, 512, 1, 512] flags = (bcast_scalar) data_type = f32
  xsmm.unary zero(data_type = f32, %0, %cst, %alloc) : (i64, f32, memref<32x32xf32>) -> ()
  %sub = memref.subview %alloc[0, 0] [16, 16] [1, 1] : memref<32x32xf32> to memref<16x16xf32, strided<[32, 1]>>
  call @test(%sub) : (memref<16x16xf32, strided<[32, 1]>>) -> ()
  // Inserts beta_0 while we cannot prove it is zero anymore.
  %1 = xsmm.brgemm.dispatch [32, 32, 32, 32, 32, 32, 32, 1] flags = (none) data_type = f32
  xsmm.brgemm(data_type = f32, %1, %arg0, %arg1, %alloc, %c1_i64) : (i64, memref<1x32x32xf32>, memref<1x32x32xf32>, memref<32x32xf32>, i64) -> ()
  return
}
func.func private @test(%arg0 : memref<16x16xf32, strided<[32, 1]>>)
```

Reply:
This still is fails if `memref.subview` is created before xsmm zero op.

Reply:
Not sure if I get it. You can zero a subview, as long as you don't do something with it before the GEMM.

Reply:
```
  %alloc = memref.alloc() {alignment = 64 : i64} : memref<32x32xf32>
  %sub = memref.subview %alloc[0, 0] [16, 16] [1, 1] : memref<32x32xf32> to memref<16x16xf32, strided<[32, 1]>>
  %0 = xsmm.unary.dispatch zero [32, 512, 1, 512] flags = (bcast_scalar) data_type = f32
  xsmm.unary zero(data_type = f32, %0, %cst, %alloc) : (i64, f32, memref<32x32xf32>) -> ()
  call @test(%sub) : (memref<16x16xf32, strided<[32, 1]>>) -> ()
  %1 = xsmm.brgemm.dispatch [32, 32, 32, 32, 32, 32, 32, 1] flags = (none) data_type = f32
```

In this pattern, folding shouldn't kick in as `%alloc` is modified through an alias.

Reply:
thanks, fixed.